<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Khaled Al-Shamaa" />

<meta name="date" content="2025-07-31" />

<title>GIGWA Example</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">GIGWA Example</h1>
<h4 class="author">Khaled Al-Shamaa</h4>
<h4 class="date">2025-07-31</h4>



<div id="qbms" class="section level2">
<h2>QBMS</h2>
<p>This R package assists breeders in linking data systems with their
analytic pipelines, a crucial step in digitizing breeding processes. It
supports querying and retrieving phenotypic and genotypic data from
systems like <a href="https://ebs.excellenceinbreeding.org/">EBS</a>, <a href="https://bmspro.io/">BMS</a>, <a href="https://breedbase.org">BreedBase</a>, and <a href="https://github.com/SouthGreenPlatform/Gigwa2">GIGWA</a> (using <a href="https://brapi.org/">BrAPI</a> calls). Extra helper functions
support environmental data sources, including <a href="https://www.climatologylab.org/terraclimate.html">TerraClimate</a>
and FAO <a href="https://gaez.fao.org/pages/hwsd">HWSDv2</a> soil
database.</p>
</div>
<div id="gigwa" class="section level2">
<h2>GIGWA</h2>
<p><a href="https://github.com/SouthGreenPlatform/Gigwa2">GIGWA</a> is a
web-based tool which provides an easy and intuitive way to explore large
amounts of genotyping data by filtering the latter based not only on
variant features, including functional annotations, but also on genotype
patterns. The data storage relies on MongoDB, which offers good
scalability perspectives. GIGWA can handle multiple databases and may be
deployed in either single or multi-user mode. Finally, it provides a
wide range of popular export formats.</p>
</div>
<div id="brapi" class="section level2">
<h2>BrAPI</h2>
<p>The Breeding API (<a href="https://brapi.org/">BrAPI</a>) project is
an effort to enable interoperability among plant breeding databases.
BrAPI is a standardized RESTful web service API specification for
communicating plant breeding data. This community driven standard is
free to be used by anyone interested in plant breeding data
management.</p>
</div>
<div id="example" class="section level2">
<h2><em>Example</em></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># load the QBMS library</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(QBMS)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># The public GIGWA testing server required no authentication. If your GIGWA server </span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># requires authentication, then make sure that no_auth parameter value is FALSE</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># IMPORTENT </span><span class="al">NOTE</span><span class="co">: QBMS required GIGWA version 2.4.1 or higher</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">set_qbms_config</span>(<span class="st">&quot;https://gigwa.southgreen.fr/gigwa/&quot;</span>, </span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>                <span class="at">time_out =</span> <span class="dv">300</span>, <span class="at">engine =</span> <span class="st">&quot;gigwa&quot;</span>, <span class="at">no_auth =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># If login is required, then you can use your GIGWA account (interactive mode)</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co"># or pass your GIGWA username and password as parameters (batch mode)</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co"># login_gigwa()</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># login_gigwa(&quot;gigwadmin&quot;, &quot;nimda&quot;)</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co"># list existing databases in the current GIGWA server</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="fu">gigwa_list_dbs</span>()</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co"># select a database by name</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="fu">gigwa_set_db</span>(<span class="st">&quot;Sorghum-JGI_v1&quot;</span>)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co"># list all projects in the selected database</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="fu">gigwa_list_projects</span>()</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co"># select a project by name</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="fu">gigwa_set_project</span>(<span class="st">&quot;Nelson_et_al_2011&quot;</span>)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co"># list all runs in the selected project</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="fu">gigwa_list_runs</span>()</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="co"># select a specific run by name</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="fu">gigwa_set_run</span>(<span class="st">&quot;run1&quot;</span>)</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="co"># get a list of all samples in the selected run</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">gigwa_get_samples</span>()</span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co"># show the first 6 individuals on the list of samples</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="fu">head</span>(samples)</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="co"># query the variants (e.g., SNPs markers) in the selected run </span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="co"># that match the given criteria:</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="co"># - max_missing: maximum missing ratio (by sample) [0-1] (default is 1 for 100%) </span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="co"># - min_maf: minimum Minor Allele Frequency (MAF) [0-1] (default is 0 for 0%) </span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="co"># - start: start position of region (zero-based, inclusive) (e.g., 19750802)</span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="co"># - end: end position of region (zero-based, exclusive) (e.g., 19850125)</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="co"># - referenceName: reference sequence name  (e.g., &#39;6H&#39; in the Barley LI-AM)</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="co"># - samples: a list of a samples subset (default is NULL will retrieve for all samples) </span></span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a>marker_matrix <span class="ot">&lt;-</span> <span class="fu">gigwa_get_variants</span>(<span class="at">max_missing =</span> <span class="fl">0.2</span>, </span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a>                                    <span class="at">min_maf =</span> <span class="fl">0.05</span>, </span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a>                                    <span class="at">start =</span> <span class="dv">100000</span>,</span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a>                                    <span class="at">end =</span> <span class="dv">500000</span>,</span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a>                                    <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;ind1&quot;</span>, <span class="st">&quot;ind3&quot;</span>, <span class="st">&quot;ind7&quot;</span>))</span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="co"># Data returns in data.frame format. The first 4 columns describe attributes of the SNP </span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="co"># - rs#: variant name</span></span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="co"># - alleles: reference allele / alternative allele</span></span>
<span id="cb1-56"><a href="#cb1-56" tabindex="-1"></a><span class="co"># - chrom: chromosome name</span></span>
<span id="cb1-57"><a href="#cb1-57" tabindex="-1"></a><span class="co"># - pos: position in bp</span></span>
<span id="cb1-58"><a href="#cb1-58" tabindex="-1"></a><span class="co"># while the following columns describe the SNP value for a single sample line using </span></span>
<span id="cb1-59"><a href="#cb1-59" tabindex="-1"></a><span class="co"># numerical coding 0, 1, and 2 for reference, heterozygous, alternative/minor alleles.</span></span>
<span id="cb1-60"><a href="#cb1-60" tabindex="-1"></a><span class="fu">head</span>(marker_matrix)</span>
<span id="cb1-61"><a href="#cb1-61" tabindex="-1"></a></span>
<span id="cb1-62"><a href="#cb1-62" tabindex="-1"></a><span class="co"># get the metadata associated with the samples in the current active run</span></span>
<span id="cb1-63"><a href="#cb1-63" tabindex="-1"></a><span class="fu">gigwa_set_db</span>(<span class="st">&quot;DIVRICE_NB&quot;</span>)</span>
<span id="cb1-64"><a href="#cb1-64" tabindex="-1"></a><span class="fu">gigwa_set_project</span>(<span class="st">&quot;refNB&quot;</span>)</span>
<span id="cb1-65"><a href="#cb1-65" tabindex="-1"></a><span class="fu">gigwa_set_run</span>(<span class="st">&quot;03052022&quot;</span>)</span>
<span id="cb1-66"><a href="#cb1-66" tabindex="-1"></a></span>
<span id="cb1-67"><a href="#cb1-67" tabindex="-1"></a><span class="co"># get a list of all samples in the selected run</span></span>
<span id="cb1-68"><a href="#cb1-68" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">gigwa_get_metadata</span>()</span>
<span id="cb1-69"><a href="#cb1-69" tabindex="-1"></a></span>
<span id="cb1-70"><a href="#cb1-70" tabindex="-1"></a><span class="fu">View</span>(metadata)</span></code></pre></div>
</div>
<div id="enhanced-allele-matrix-retrieval" class="section level2">
<h2>Enhanced Allele Matrix Retrieval</h2>
<p>The following functions now utilize the new efficient BrAPI v2.1
/allelematrix calls, requiring version 2.6 of GIGWA or higher. This
update significantly improves the QBMS allele matrix retrieval speed,
increasing it by more than 10 times, as demonstrated by benchmark
tests.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Configure your GIGWA connection</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">set_qbms_config</span>(<span class="st">&quot;https://gigwa-dev.southgreen.fr/gigwaV2/&quot;</span>, <span class="at">page_size =</span> <span class="dv">50000</span>,</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>                <span class="at">time_out =</span> <span class="dv">300</span>, <span class="at">engine =</span> <span class="st">&quot;gigwa&quot;</span>, <span class="at">no_auth =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># Select a database by name</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">gigwa_set_db</span>(<span class="st">&quot;Sorghum-JGI_v1&quot;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Select a project by name</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">gigwa_set_project</span>(<span class="st">&quot;Nelson_et_al_2011&quot;</span>)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># Select a specific run by name</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="fu">gigwa_set_run</span>(<span class="st">&quot;run1&quot;</span>)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co"># Get the list of all samples in the selected project</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>germplasmNames <span class="ot">&lt;-</span> <span class="fu">gigwa_get_samples</span>()</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co"># Get the list of all sequences in the selected project</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>chroms <span class="ot">&lt;-</span> <span class="fu">gigwa_get_sequences</span>()</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="do">### Get Variants Info (Geno Map) ###############################################</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a>?gigwa_get_markers</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>geno_map <span class="ot">&lt;-</span> <span class="fu">gigwa_get_markers</span>(<span class="at">start =</span> <span class="dv">0</span>,</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a>                              <span class="at">end =</span> <span class="dv">1234567</span>,</span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a>                              <span class="co"># chrom = c(&quot;Sb01&quot;, &quot;Sb02&quot;),      # chroms[1:3]</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a>                              )</span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="do">### Get Marker Matrix ##########################################################</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a>?gigwa_get_allelematrix</span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a>geno_data <span class="ot">&lt;-</span> <span class="fu">gigwa_get_allelematrix</span>(<span class="at">start =</span> <span class="dv">0</span>,                  <span class="co"># default is 0</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a>                                    <span class="at">end =</span> <span class="dv">1234567</span>,              <span class="co"># default is &quot;&quot; -&gt; &quot;ref:0-&quot;</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a>                                    <span class="at">snps =</span> geno_map<span class="sc">$</span><span class="st">`</span><span class="at">rs#</span><span class="st">`</span>,      <span class="co"># optional</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a>                                    <span class="co"># chrom = &quot;Sb01&quot;,           # c(&quot;Sb01&quot;, &quot;Sb07&quot;)</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a>                                    <span class="co"># samples = germplasmNames, # gigwa_get_samples()</span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a>                                    <span class="co"># snps_pageSize = 10000,</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a>                                    <span class="co"># samples_pageSize = 100,</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a>                                    )</span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a><span class="do">### Get The Whole Variant Set ##################################################</span></span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a></span>
<span id="cb2-42"><a href="#cb2-42" tabindex="-1"></a>?get_variantset</span>
<span id="cb2-43"><a href="#cb2-43" tabindex="-1"></a>snps <span class="ot">&lt;-</span> <span class="fu">get_variantset</span>()</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
